File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-whole-object-default-losing-per-key-defaults.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findProgramRoot } from "../../utils/find-program-root.js";
import { findTransparentExpressionRoot } from "../../utils/find-transparent-expression-root.js";
import { isAstNode } from "../../utils/is-ast-node.js";
import { isFunctionLike } from "../../utils/is-function-like.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { RuleVisitors } from "../../utils/rule-visitors.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import { walkAst } from "../../utils/walk-ast.js";

// A whole-object parameter default (`({ a, b } = { a: 1, b: 'x' })`)
// only applies when the argument is omitted ENTIRELY. The instant a
// caller passes any object, the default object is discarded wholesale
// and every key the caller left out becomes `undefined` — silently
// bypassing the intended fallback. The correct form applies each
// default independently: `({ a = 1, b = 'x' } = {})`.
//
// Scope-fix (fires only when a per-key default is actually reachable to
// lose, and losing it changes behavior):
//   1. the parameter is `ObjectPattern = ObjectExpression` (TS wrappers
//      like `as` / `satisfies` around the object are peeled first),
//   2. at least one destructured binding lacks its OWN default AND has a
//      matching `key: value` property in the default object — that value
//      is the fallback a partial call silently drops,
//   3. at least one such dropped fallback is something other than the
//      literal `false` or `undefined` — boolean-flag bags defaulting
//      everything to `false` are consumed in truthiness checks where
//      `undefined` behaves identically, and an `undefined` fallback IS
//      the value the key would get anyway, so nothing observable is
//      lost either way,
//   4. the function has an unknown caller: when it is a local,
//      non-exported binding whose every in-file reference is a direct
//      call that either omits the argument or passes an object literal
//      covering every at-risk key, no caller can ever apply the default
//      partially, so the hazard is unreachable.
// `= {}` (the recommended idiom), patterns where every binding is
// already defaulted, all-`false`/`undefined` flag bags, and locals
// whose call sites all pass complete objects stay quiet.

const getStaticPropertyKey = (property: EsTreeNodeOfType<"Property">): string | null => {
  if (property.computed) return null;
  const key = property.key as EsTreeNode;
  if (isNodeOfType(key, "Identifier")) return key.name;
  if (isNodeOfType(key, "Literal")) return String(key.value);
  return null;
};

// Keys of bindings that carry no `= default` of their own — the bindings
// whose fallback the whole-object default silently drops on a
// partial-argument call.
const collectUndefaultedBindingKeys = (
  objectPattern: EsTreeNodeOfType<"ObjectPattern">,
): Set<string> => {
  const undefaultedBindingKeys = new Set<string>();
  for (const property of objectPattern.properties ?? []) {
    if (!isNodeOfType(property, "Property")) continue;
    if (isNodeOfType(property.value as EsTreeNode, "AssignmentPattern")) continue;
    const bindingKey = getStaticPropertyKey(property);
    if (bindingKey !== null) undefaultedBindingKeys.add(bindingKey);
  }
  return undefaultedBindingKeys;
};

interface DroppedFallback {
  key: string;
  value: EsTreeNode;
}

// The default-object values that a partial call actually drops: each
// `key: value` whose key matches an undefaulted binding.
const collectDroppedFallbacks = (
  objectExpression: EsTreeNodeOfType<"ObjectExpression">,
  undefaultedBindingKeys: Set<string>,
): Array<DroppedFallback> => {
  const droppedFallbacks: Array<DroppedFallback> = [];
  for (const property of objectExpression.properties ?? []) {
    if (!isNodeOfType(property, "Property")) continue;
    const propertyKey = getStaticPropertyKey(property);
    if (propertyKey === null || !undefaultedBindingKeys.has(propertyKey)) continue;
    droppedFallbacks.push({ key: propertyKey, value: property.value as EsTreeNode });
  }
  return droppedFallbacks;
};

// `false` fallbacks feed truthiness checks where `undefined` behaves
// identically, so dropping them is a runtime no-op.
const isFalseLiteral = (value: EsTreeNode): boolean => {
  const innerValue = stripParenExpression(value);
  return isNodeOfType(innerValue, "Literal") && innerValue.value === false;
};

// An `undefined` fallback IS what the key resolves to when dropped, so
// losing it changes nothing.
const isUndefinedValue = (value: EsTreeNode): boolean => {
  const innerValue = stripParenExpression(value);
  if (isNodeOfType(innerValue, "Identifier")) return innerValue.name === "undefined";
  return isNodeOfType(innerValue, "UnaryExpression") && innerValue.operator === "void";
};

const getPatternBindingForKey = (
  objectPattern: EsTreeNodeOfType<"ObjectPattern">,
  key: string,
): EsTreeNodeOfType<"Identifier"> | null => {
  for (const property of objectPattern.properties ?? []) {
    if (!isNodeOfType(property, "Property") || getStaticPropertyKey(property) !== key) continue;
    if (isNodeOfType(property.value, "Identifier")) return property.value;
  }
  return null;
};

const isNoOpFunctionValue = (value: EsTreeNode): boolean => {
  const unwrappedValue = stripParenExpression(value);
  if (!isFunctionLike(unwrappedValue)) return false;
  if (!isNodeOfType(unwrappedValue.body, "BlockStatement")) {
    return Boolean(unwrappedValue.body && isUndefinedValue(unwrappedValue.body));
  }
  return unwrappedValue.body.body.every(
    (statement) =>
      isNodeOfType(statement, "EmptyStatement") ||
      (isNodeOfType(statement, "ReturnStatement") && !statement.argument),
  );
};

const bindingIsOnlyOptionallyCalled = (
  binding: EsTreeNodeOfType<"Identifier">,
  context: RuleContext,
): boolean => {
  const symbol = context.scopes.symbolFor(binding);
  if (!symbol) return false;
  return symbol.references.every((reference) => {
    const referenceRoot = findTransparentExpressionRoot(reference.identifier);
    const parent = referenceRoot.parent;
    return Boolean(
      parent &&
      isNodeOfType(parent, "CallExpression") &&
      parent.callee === referenceRoot &&
      parent.optional === true,
    );
  });
};

const isNoOpOptionalCallbackFallback = (
  fallback: DroppedFallback,
  objectPattern: EsTreeNodeOfType<"ObjectPattern">,
  context: RuleContext,
): boolean => {
  if (!isNoOpFunctionValue(fallback.value)) return false;
  const binding = getPatternBindingForKey(objectPattern, fallback.key);
  return binding !== null && bindingIsOnlyOptionallyCalled(binding, context);
};

const getTypeMemberKey = (member: EsTreeNode): string | null => {
  if (!isNodeOfType(member, "TSPropertySignature") && !isNodeOfType(member, "TSMethodSignature")) {
    return null;
  }
  if (member.computed) return null;
  if (isNodeOfType(member.key, "Identifier")) return member.key.name;
  if (isNodeOfType(member.key, "Literal")) return String(member.key.value);
  return null;
};

const isRequiredTypeMember = (member: EsTreeNode, propertyName: string): boolean => {
  if (!isNodeOfType(member, "TSPropertySignature") && !isNodeOfType(member, "TSMethodSignature")) {
    return false;
  }
  return getTypeMemberKey(member) === propertyName && member.optional !== true;
};

const findSameFileTypeDeclaration = (
  referenceNode: EsTreeNode,
  typeName: string,
): EsTreeNode | null => {
  const programRoot = findProgramRoot(referenceNode);
  if (!programRoot) return null;
  const declarations: EsTreeNode[] = [];
  walkAst(programRoot, (candidate: EsTreeNode) => {
    if (
      (isNodeOfType(candidate, "TSInterfaceDeclaration") ||
        isNodeOfType(candidate, "TSTypeAliasDeclaration")) &&
      candidate.id.name === typeName
    ) {
      declarations.push(candidate);
      return false;
    }
  });
  return declarations.length === 1 ? declarations[0] : null;
};

const typeProvesRequiredProperty = (
  typeNode: EsTreeNode,
  propertyName: string,
  referenceNode: EsTreeNode,
  visitedDeclarations: Set<EsTreeNode>,
): boolean => {
  if (isNodeOfType(typeNode, "TSTypeAnnotation")) {
    return typeProvesRequiredProperty(
      typeNode.typeAnnotation,
      propertyName,
      referenceNode,
      visitedDeclarations,
    );
  }
  if (isNodeOfType(typeNode, "TSTypeLiteral")) {
    return typeNode.members.some((member) => isRequiredTypeMember(member, propertyName));
  }
  if (isNodeOfType(typeNode, "TSInterfaceDeclaration")) {
    return typeNode.body.body.some((member) => isRequiredTypeMember(member, propertyName));
  }
  if (isNodeOfType(typeNode, "TSIntersectionType")) {
    return typeNode.types.some((intersectionMember) =>
      typeProvesRequiredProperty(
        intersectionMember,
        propertyName,
        referenceNode,
        visitedDeclarations,
      ),
    );
  }
  if (isNodeOfType(typeNode, "TSTypeAliasDeclaration")) {
    if (visitedDeclarations.has(typeNode)) return false;
    visitedDeclarations.add(typeNode);
    const isRequired = typeProvesRequiredProperty(
      typeNode.typeAnnotation,
      propertyName,
      referenceNode,
      visitedDeclarations,
    );
    visitedDeclarations.delete(typeNode);
    return isRequired;
  }
  if (
    !isNodeOfType(typeNode, "TSTypeReference") ||
    !isNodeOfType(typeNode.typeName, "Identifier")
  ) {
    return false;
  }
  const declaration = findSameFileTypeDeclaration(referenceNode, typeNode.typeName.name);
  if (!declaration || visitedDeclarations.has(declaration)) return false;
  visitedDeclarations.add(declaration);
  const isRequired = typeProvesRequiredProperty(
    declaration,
    propertyName,
    referenceNode,
    visitedDeclarations,
  );
  visitedDeclarations.delete(declaration);
  return isRequired;
};

const patternTypeRequiresEveryFallback = (
  objectPattern: EsTreeNodeOfType<"ObjectPattern">,
  fallbacks: ReadonlyArray<DroppedFallback>,
): boolean => {
  const annotation = objectPattern.typeAnnotation;
  if (!annotation) return false;
  return fallbacks.every((fallback) =>
    typeProvesRequiredProperty(annotation, fallback.key, objectPattern, new Set()),
  );
};

// True when the AssignmentPattern is a direct parameter of a function
// (not a nested destructuring default inside another pattern).
const isFunctionParameter = (assignmentPattern: EsTreeNode): boolean => {
  const parent = assignmentPattern.parent;
  return Boolean(
    parent &&
    isFunctionLike(parent) &&
    parent.params?.some((parameter) => parameter === assignmentPattern),
  );
};

interface LocalFunctionBinding {
  name: string;
  bindingIdentifier: EsTreeNode;
}

// The identifier a plain `function f` declaration or `const f = () => {}`
// declarator binds the function to — null for inline callbacks, methods,
// and any exported form (an export means callers we cannot see).
const resolveUnexportedFunctionBinding = (
  functionNode: EsTreeNode,
): LocalFunctionBinding | null => {
  if (isNodeOfType(functionNode, "FunctionDeclaration") && functionNode.id) {
    const declarationParent = functionNode.parent;
    if (declarationParent && declarationParent.type.startsWith("Export")) return null;
    return { name: functionNode.id.name, bindingIdentifier: functionNode.id as EsTreeNode };
  }
  const declarator = functionNode.parent;
  if (!declarator || !isNodeOfType(declarator, "VariableDeclarator")) return null;
  if (!isNodeOfType(declarator.id, "Identifier")) return null;
  const declaration = declarator.parent;
  if (declaration?.parent && declaration.parent.type.startsWith("Export")) return null;
  return { name: declarator.id.name, bindingIdentifier: declarator.id as EsTreeNode };
};

// A spread-free object-literal argument statically providing every
// at-risk key — such a call can never trigger the dropped fallbacks.
const callArgumentCoversKeys = (argument: EsTreeNode, atRiskKeys: Set<string>): boolean => {
  const stripped = stripParenExpression(argument);
  if (!isNodeOfType(stripped, "ObjectExpression")) return false;
  const providedKeys = new Set<string>();
  for (const property of stripped.properties ?? []) {
    if (!isNodeOfType(property, "Property")) return false;
    const propertyKey = getStaticPropertyKey(property);
    if (propertyKey !== null) providedKeys.add(propertyKey);
  }
  return [...atRiskKeys].every((atRiskKey) => providedKeys.has(atRiskKey));
};

// Whole-file reference scan: true only when the function has at least one
// in-file call and EVERY reference is a direct call that either omits the
// argument entirely (the whole-object default applies intact) or passes an
// object literal covering every at-risk key. Any bare/JSX/exported use means
// unknown callers, so the guard declines.
const everyCallSitePassesAtRiskKeys = (
  functionNode: EsTreeNode,
  parameterIndex: number,
  atRiskKeys: Set<string>,
): boolean => {
  const binding = resolveUnexportedFunctionBinding(functionNode);
  if (!binding) return false;
  const programRoot = findProgramRoot(functionNode);
  if (!programRoot) return false;
  let sawCompleteCall = false;
  let sawUnknowableReference = false;
  walkAst(programRoot, (child: EsTreeNode) => {
    if (sawUnknowableReference) return false;
    if (isNodeOfType(child, "JSXIdentifier") && child.name === binding.name) {
      sawUnknowableReference = true;
      return false;
    }
    if (!isNodeOfType(child, "Identifier") || child.name !== binding.name) return;
    if (child === binding.bindingIdentifier) return;
    const parent = child.parent;
    if (!parent) return;
    if (isNodeOfType(parent, "MemberExpression") && parent.property === child && !parent.computed) {
      return;
    }
    if (
      isNodeOfType(parent, "Property") &&
      parent.key === child &&
      !parent.computed &&
      !parent.shorthand
    ) {
      return;
    }
    if (isNodeOfType(parent, "CallExpression") && parent.callee === child) {
      const argument = parent.arguments?.[parameterIndex];
      if (!isAstNode(argument) || callArgumentCoversKeys(argument, atRiskKeys)) {
        sawCompleteCall = true;
        return;
      }
    }
    sawUnknowableReference = true;
    return false;
  });
  return sawCompleteCall && !sawUnknowableReference;
};

export const noWholeObjectDefaultLosingPerKeyDefaults = defineRule({
  id: "no-whole-object-default-losing-per-key-defaults",
  title: "Whole-object param default loses per-key defaults",
  severity: "warn",
  category: "Correctness",
  tags: ["test-noise"],
  recommendation:
    "A whole-object parameter default applies only when the argument is omitted entirely, so a partial argument makes every omitted key undefined. Move each fallback onto its own binding instead: `({ a = 1, b = false } = {})`.",
  create: (context: RuleContext): RuleVisitors => ({
    AssignmentPattern(node: EsTreeNodeOfType<"AssignmentPattern">) {
      if (!isFunctionParameter(node)) return;
      const pattern = node.left as EsTreeNode;
      const defaultValue = stripParenExpression(node.right as EsTreeNode);
      if (!isNodeOfType(pattern, "ObjectPattern")) return;
      if (!isNodeOfType(defaultValue, "ObjectExpression")) return;
      const undefaultedBindingKeys = collectUndefaultedBindingKeys(pattern);
      if (undefaultedBindingKeys.size === 0) return;
      const droppedFallbacks = collectDroppedFallbacks(defaultValue, undefaultedBindingKeys);
      if (droppedFallbacks.length === 0) return;
      const observableDroppedFallbacks = droppedFallbacks.filter(
        (fallback) =>
          !isFalseLiteral(fallback.value) &&
          !isUndefinedValue(fallback.value) &&
          !isNoOpOptionalCallbackFallback(fallback, pattern, context),
      );
      if (observableDroppedFallbacks.length === 0) return;
      if (patternTypeRequiresEveryFallback(pattern, observableDroppedFallbacks)) return;
      const enclosingFunction = node.parent;
      if (enclosingFunction && isFunctionLike(enclosingFunction)) {
        const parameterIndex = (enclosingFunction.params ?? []).findIndex(
          (parameter) => parameter === node,
        );
        const atRiskKeys = new Set(observableDroppedFallbacks.map((fallback) => fallback.key));
        if (
          parameterIndex >= 0 &&
          everyCallSitePassesAtRiskKeys(enclosingFunction, parameterIndex, atRiskKeys)
        ) {
          return;
        }
      }
      context.report({
        node,
        message:
          "This whole-object default is discarded the moment a caller passes any object, so every omitted key becomes undefined instead of falling back. Give each binding its own default instead: `({ a = 1, b = false } = {})`.",
      });
    },
  }),
});

File: native/oxlint/rules/no-whole-object-default-losing-per-key-defaults.rs
use oxc_ast::{
    ast::{
        Argument, BindingPattern, Declaration, Expression, ObjectPropertyKind, Statement,
        TSSignature, TSType, TSTypeName,
    },
    AstKind,
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_semantic::{NodeId, SymbolId};
use oxc_span::{GetSpan, Span};
use oxc_syntax::operator::UnaryOperator;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::{
    context::{ContextHost, LintContext},
    rule::Rule,
};

const MESSAGE: &str = "This whole-object default is discarded the moment a caller passes any object, so every omitted key becomes undefined instead of falling back. Give each binding its own default instead: `({ a = 1, b = false } = {})`.";

#[derive(Debug, Default, Clone)]
pub struct NoWholeObjectDefaultLosingPerKeyDefaults;

declare_oxc_lint!(
    /// Warns when a whole-object parameter default loses per-key fallbacks.
    NoWholeObjectDefaultLosingPerKeyDefaults,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Whole-object parameter default loses per-key defaults.",
);

impl Rule for NoWholeObjectDefaultLosingPerKeyDefaults {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }

    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        let type_index = whole_default_build_type_index(ctx);
        for node in ctx.nodes().iter() {
            let parameters = match node.kind() {
                AstKind::Function(function) => &function.params,
                AstKind::ArrowFunctionExpression(function) => &function.params,
                _ => continue,
            };
            for (parameter_index, parameter) in parameters.items.iter().enumerate() {
                let BindingPattern::ObjectPattern(object_pattern) = &parameter.pattern else {
                    continue;
                };
                let Some(initializer) = parameter.initializer.as_ref() else {
                    continue;
                };
                let Expression::ObjectExpression(default_object) =
                    initializer.get_inner_expression()
                else {
                    continue;
                };

                let undefaulted_keys = object_pattern
                    .properties
                    .iter()
                    .filter(|property| {
                        !property.computed
                            && !matches!(property.value, BindingPattern::AssignmentPattern(_))
                    })
                    .filter_map(|property| property.key.static_name().map(|name| name.to_string()))
                    .collect::<FxHashSet<_>>();
                if undefaulted_keys.is_empty() {
                    continue;
                }

                let observable_fallbacks = default_object
                    .properties
                    .iter()
                    .filter_map(|property| {
                        let ObjectPropertyKind::ObjectProperty(property) = property else {
                            return None;
                        };
                        if property.computed {
                            return None;
                        }
                        let key = property.key.static_name()?.to_string();
                        undefaulted_keys
                            .contains(&key)
                            .then_some((key, &property.value))
                    })
                    .filter(|(key, value)| {
                        !whole_default_is_undefined_value(value, ctx)
                            && !whole_default_is_noop_optional_callback(
                                object_pattern,
                                key,
                                value,
                                ctx,
                            )
                    })
                    .collect::<Vec<_>>();
                if observable_fallbacks.is_empty() {
                    continue;
                }
                if parameter
                    .type_annotation
                    .as_ref()
                    .is_some_and(|annotation| {
                        observable_fallbacks.iter().all(|(key, _)| {
                            whole_default_type_requires_property(
                                WholeDefaultType::Type(&annotation.type_annotation),
                                key,
                                &type_index,
                                &mut FxHashSet::default(),
                            )
                        })
                    })
                {
                    continue;
                }

                let at_risk_keys = observable_fallbacks
                    .iter()
                    .map(|(key, _)| key.clone())
                    .collect::<FxHashSet<_>>();
                if whole_default_every_call_site_covers_keys(
                    node.id(),
                    parameter_index,
                    &at_risk_keys,
                    ctx,
                ) {
                    continue;
                }
                ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(Span::new(
                    parameter.pattern.span().start,
                    initializer.span().end,
                )));
            }
        }
    }
}

#[derive(Clone, Copy)]
enum WholeDefaultType<'a> {
    Type(&'a TSType<'a>),
    Interface(&'a oxc_ast::ast::TSInterfaceDeclaration<'a>),
    Alias(&'a oxc_ast::ast::TSTypeAliasDeclaration<'a>),
}

type WholeDefaultTypeIndex<'a> = FxHashMap<String, Vec<WholeDefaultType<'a>>>;

fn whole_default_build_type_index<'a>(ctx: &LintContext<'a>) -> WholeDefaultTypeIndex<'a> {
    let mut type_index = WholeDefaultTypeIndex::default();
    let Some(program) = ctx.nodes().iter().find_map(|node| match node.kind() {
        AstKind::Program(program) => Some(program),
        _ => None,
    }) else {
        return type_index;
    };
    for statement in &program.body {
        let declaration = match statement {
            Statement::TSInterfaceDeclaration(interface) => WholeDefaultType::Interface(interface),
            Statement::TSTypeAliasDeclaration(alias) => WholeDefaultType::Alias(alias),
            Statement::ExportDeclaration(export) => match &export.declaration {
                Declaration::TSInterfaceDeclaration(interface) => {
                    WholeDefaultType::Interface(interface)
                }
                Declaration::TSTypeAliasDeclaration(alias) => WholeDefaultType::Alias(alias),
                _ => continue,
            },
            _ => continue,
        };
        let name = match declaration {
            WholeDefaultType::Interface(interface) => interface.id.name.to_string(),
            WholeDefaultType::Alias(alias) => alias.id.name.to_string(),
            WholeDefaultType::Type(_) => continue,
        };
        type_index.entry(name).or_default().push(declaration);
    }
    type_index
}

fn whole_default_is_undefined_value(expression: &Expression<'_>, ctx: &LintContext<'_>) -> bool {
    match expression.get_inner_expression() {
        Expression::Identifier(identifier) => {
            identifier.name == "undefined"
                && ctx
                    .scoping()
                    .get_reference(identifier.reference_id())
                    .symbol_id()
                    .is_none()
        }
        Expression::UnaryExpression(unary) => unary.operator == UnaryOperator::Void,
        _ => false,
    }
}

fn whole_default_is_noop_optional_callback(
    object_pattern: &oxc_ast::ast::ObjectPattern<'_>,
    fallback_key: &str,
    fallback_value: &Expression<'_>,
    ctx: &LintContext<'_>,
) -> bool {
    if !whole_default_is_noop_function(fallback_value, ctx) {
        return false;
    }
    let Some(binding_symbol_id) = object_pattern.properties.iter().find_map(|property| {
        if property.computed || property.key.static_name().as_deref() != Some(fallback_key) {
            return None;
        }
        let BindingPattern::BindingIdentifier(binding) = &property.value else {
            return None;
        };
        Some(binding.symbol_id())
    }) else {
        return false;
    };
    ctx.scoping()
        .get_resolved_references(binding_symbol_id)
        .all(|reference| {
            let reference_node = ctx.nodes().get_node(reference.node_id());
            let reference_root = transparent_expression_root(reference_node, ctx);
            matches!(ctx.nodes().parent_node(reference_root.id()).kind(), AstKind::CallExpression(call)
                if call.optional && call.callee.span() == reference_root.span())
        })
}

fn whole_default_is_noop_function(expression: &Expression<'_>, ctx: &LintContext<'_>) -> bool {
    match expression.get_inner_expression() {
        Expression::ArrowFunctionExpression(function) => {
            if let Some(body_expression) = function.get_expression() {
                return whole_default_is_undefined_value(body_expression, ctx);
            }
            let oxc_ast::ast::ArrowFunctionBody::FunctionBody(body) = &function.body else {
                return false;
            };
            whole_default_statements_are_noop(&body.statements)
        }
        Expression::FunctionExpression(function) => function
            .body
            .as_ref()
            .is_some_and(|body| whole_default_statements_are_noop(&body.statements)),
        _ => false,
    }
}

fn whole_default_statements_are_noop(statements: &[oxc_ast::ast::Statement<'_>]) -> bool {
    statements.iter().all(|statement| {
        matches!(statement, oxc_ast::ast::Statement::EmptyStatement(_))
            || matches!(statement, oxc_ast::ast::Statement::ReturnStatement(return_statement)
                if return_statement.argument.is_none())
    })
}

fn whole_default_type_requires_property(
    resolved: WholeDefaultType<'_>,
    property_name: &str,
    type_index: &WholeDefaultTypeIndex<'_>,
    visited_declaration_starts: &mut FxHashSet<u32>,
) -> bool {
    match resolved {
        WholeDefaultType::Alias(alias) => {
            if !visited_declaration_starts.insert(alias.span.start) {
                return false;
            }
            let result = whole_default_type_requires_property(
                WholeDefaultType::Type(&alias.type_annotation),
                property_name,
                type_index,
                visited_declaration_starts,
            );
            visited_declaration_starts.remove(&alias.span.start);
            result
        }
        WholeDefaultType::Interface(interface) => {
            if !visited_declaration_starts.insert(interface.span.start) {
                return false;
            }
            let result =
                whole_default_members_require_property(&interface.body.body, property_name)
                    || interface.extends.iter().any(|heritage| {
                        if heritage.type_arguments.is_some() {
                            return false;
                        }
                        let Some(type_name) = whole_default_type_name(&heritage.type_name) else {
                            return false;
                        };
                        type_index.get(type_name).is_some_and(|declarations| {
                            declarations.iter().copied().any(|declaration| {
                                whole_default_type_requires_property(
                                    declaration,
                                    property_name,
                                    type_index,
                                    visited_declaration_starts,
                                )
                            })
                        })
                    });
            visited_declaration_starts.remove(&interface.span.start);
            result
        }
        WholeDefaultType::Type(type_node) => match type_node {
            TSType::TSTypeLiteral(literal) => {
                whole_default_members_require_property(&literal.members, property_name)
            }
            TSType::TSIntersectionType(intersection) => intersection.types.iter().any(|member| {
                whole_default_type_requires_property(
                    WholeDefaultType::Type(member),
                    property_name,
                    type_index,
                    visited_declaration_starts,
                )
            }),
            TSType::TSUnionType(union) => union.types.iter().all(|member| {
                whole_default_type_requires_property(
                    WholeDefaultType::Type(member),
                    property_name,
                    type_index,
                    visited_declaration_starts,
                )
            }),
            TSType::TSTypeReference(reference) => {
                let Some(type_name) = whole_default_type_name(&reference.type_name) else {
                    return false;
                };
                type_index.get(type_name).is_some_and(|declarations| {
                    declarations.iter().copied().any(|declaration| {
                        whole_default_type_requires_property(
                            declaration,
                            property_name,
                            type_index,
                            visited_declaration_starts,
                        )
                    })
                })
            }
            _ => false,
        },
    }
}

fn whole_default_type_name<'a>(type_name: &'a TSTypeName<'a>) -> Option<&'a str> {
    let TSTypeName::IdentifierReference(identifier) = type_name else {
        return None;
    };
    Some(identifier.name.as_str())
}

fn whole_default_members_require_property(
    members: &[TSSignature<'_>],
    property_name: &str,
) -> bool {
    members.iter().any(|member| match member {
        TSSignature::TSPropertySignature(property) => {
            !property.computed
                && property.key.static_name().as_deref() == Some(property_name)
                && !property.optional
        }
        TSSignature::TSMethodSignature(method) => {
            !method.computed
                && method.key.static_name().as_deref() == Some(property_name)
                && !method.optional
        }
        _ => false,
    })
}

fn whole_default_every_call_site_covers_keys(
    function_id: NodeId,
    parameter_index: usize,
    at_risk_keys: &FxHashSet<String>,
    ctx: &LintContext<'_>,
) -> bool {
    let Some(function_symbol_id) = whole_default_unexported_function_symbol(function_id, ctx)
    else {
        return false;
    };
    let mut saw_complete_call = false;
    for reference in ctx.scoping().get_resolved_references(function_symbol_id) {
        let reference_node = ctx.nodes().get_node(reference.node_id());
        let parent = ctx.nodes().parent_node(reference_node.id());
        if matches!(parent.kind(), AstKind::StaticMemberExpression(member)
            if member.property.span == reference_node.span())
        {
            continue;
        }
        if matches!(parent.kind(), AstKind::ObjectProperty(property)
            if !property.computed && !property.shorthand && property.key.span() == reference_node.span())
        {
            continue;
        }
        let AstKind::CallExpression(call) = parent.kind() else {
            return false;
        };
        if call.callee.span() != reference_node.span() {
            return false;
        }
        let argument = call.arguments.get(parameter_index);
        if argument.is_some_and(|argument| {
            !whole_default_call_argument_covers_keys(argument, at_risk_keys, ctx)
        }) {
            return false;
        }
        saw_complete_call = true;
    }
    saw_complete_call
}

fn whole_default_unexported_function_symbol(
    function_id: NodeId,
    ctx: &LintContext<'_>,
) -> Option<SymbolId> {
    let function_node = ctx.nodes().get_node(function_id);
    match function_node.kind() {
        AstKind::Function(_) | AstKind::ArrowFunctionExpression(_) => {
            let declarator_node = ctx.nodes().parent_node(function_id);
            let AstKind::VariableDeclarator(declarator) = declarator_node.kind() else {
                return None;
            };
            let binding = declarator.id.get_binding_identifier()?;
            let declaration_node = ctx.nodes().parent_node(declarator_node.id());
            if !matches!(declaration_node.kind(), AstKind::VariableDeclaration(_))
                || matches!(
                    ctx.nodes().parent_node(declaration_node.id()).kind(),
                    AstKind::ExportNamedDeclaration(_) | AstKind::ExportDefaultDeclaration(_)
                )
            {
                return None;
            }
            Some(binding.symbol_id())
        }
        _ => None,
    }
}

fn whole_default_call_argument_covers_keys(
    argument: &Argument<'_>,
    at_risk_keys: &FxHashSet<String>,
    ctx: &LintContext<'_>,
) -> bool {
    let Some(expression) = argument.as_expression() else {
        return false;
    };
    if whole_default_is_undefined_value(expression, ctx) {
        return true;
    }
    let Expression::ObjectExpression(object) = expression.get_inner_expression() else {
        return false;
    };
    let mut provided_keys = FxHashSet::default();
    for property in &object.properties {
        let ObjectPropertyKind::ObjectProperty(property) = property else {
            return false;
        };
        if property.computed {
            continue;
        }
        if let Some(property_name) = property.key.static_name() {
            provided_keys.insert(property_name.to_string());
        }
    }
    at_risk_keys.is_subset(&provided_keys)
}
